z-theory/generate_z_events.cpp

```diff
@@ -1,61 +1,30 @@
-// main92.cc is a part of the PYTHIA event generator.
-// Copyright (C) 2020 Torbjorn Sjostrand.
-// PYTHIA is licenced under the GNU GPL v2 or later, see COPYING for details.
-// Please respect the MCnet Guidelines, see GUIDELINES for details.
-
-// Keywords: analysis; root;
-
-// This is a simple test program.
-// Modified by Rene Brun and Axel Naumann to put the Pythia::event
-// into a TTree.
-
-// Header file to access Pythia 8 program elements.
 #include "Pythia8/Pythia.h"
 
-// ROOT, for saving Pythia events as trees in a file.
-#include "TTree.h"
-#include "TFile.h"
-
 using namespace Pythia8;
 
-int main() {
-
-  // Create Pythia instance and set it up to generate hard QCD processes
-  // above pTHat = 20 GeV for pp collisions at 14 TeV.
+int main()
+{
   Pythia pythia;
-  pythia.readString("HardQCD:all = on");
+  pythia.readString("WeakSingleBoson:all = on");
   pythia.readString("PhaseSpace:pTHatMin = 20.");
-  pythia.readString("Beams:eCM = 14000.");
+  pythia.readString("Beams:eCM = 7000.");
   pythia.init();
 
-  // Set up the ROOT TFile and TTree.
-  TFile *file = TFile::Open("pytree.root","recreate");
-  Event *event = &pythia.event;
-  TTree *T = new TTree("T","ev1 Tree");
-  T->Branch("event",&event);
+  
+  Event& event = pythia.event;
```

Comment:
Writing the entire pythia "event" to the TTree is probably not manageable for any serious number of events.

You can try writing single precision (32 bit) floats for the relevant variables like the mup_PT,mup_ETA,mup_PHI etc...
Something like
```cpp
float mup_PT;
T->Branch("mup_PT",&mup_PT);
for (int iEvent = 0;....){
mup_PT = 1.0;
T->Fill();
}
```

Then you need to figure out how to actually find the muons from Z in pythia....

Reply:
We could try to find the "Z" particle from the event history but I'd first suggest to try and find the two highest pT final state muons (of opposite charge).
Inside each event you can do
```cpp
const auto & event = pythia.event ;
 for ( auto i_particle = 0; i_particle < event.size(); ++i_particle ) {
    auto pT = event[i_particle].pT(); 
 auto ID = event[i_particle].pid(); 
}
```
Then you should be able to write some logic to find the highest pT final state mu+ and mu-.
Or more useful is to find their indices.
Then you can just ask for whatever property for the particles at those indices and copy their values to the TTree output variables.